### PR TITLE
LoRA Logic Bug Report:mm_tunable_parts + LoRA = Full-Model Training (Unintended)

### LINKED ISSUE #508: LoRA Logic Bug Report:mm_tunable_parts + LoRA = Full-Model Training (Unintended)
> **Preface**
> Thanks for maintaining this project. I’m sharing my findings in good faith while reproducing EgoBlind. If I’ve misunderstood your original intent, design choices, or any part of the code logic, please forgive me—and I’d really appreciate any correction. I’m happy to adjust the report/PR accordingly once I understand the intended behavior.
### LoRA’s wrong logic in LLaVA-NeXT

When I was running the [EgoBlind](https://github.com/doc-doc/EgoBlind) dataset on the LLaVA-NeXT model (which needs to train the MLP layer), I found that after using LoRA fine-tuning, the overall VRAM usage (about 160G) and the inference speed almost did not change at all! Previously, someone in the issues of your project’s GitHub repository mentioned a similar problem [issues#253](https://github.com/LLaVA-VL/LLaVA-NeXT/issues/253). This is obviously inconsistent with the general phenomenon after using LoRA fine-tuning. By reading through all relevant code in the LLaVA-NeXT repository, I found a major error in the LoRA logic in LLaVA-NeXT! Its specific manifestation is: as long as the original train.py code uses the mm_tunable_parts parameter, it is impossible to perform LoRA fine-tuning in the true sense (i.e., freezing the backbone model and only fine-tuning the LoRA parameters)!!! And if you do not use the mm_tunable_parts parameter, it is forever impossible to achieve true LoRA fine-tuning + training the MLP layer (or vision_resampler, vision_tower)! Specifically as follows:

#### Using the mm_tunable_parts parameter

First, it should be clarified that if you want to use LoRA fine-tuning, inevitably the lora_enable parameter will be set to True, and LoRA will be injected into the model first.

If you use the `mm_tunable_parts` parameter, naturally you will enter the else branch of this code block (line 1615) in trian.py:

```python
if model_args.mm_tunable_parts is None:  # traditional way of deciding which part to train
    model.config.tune_mm_mlp_adapter = training_args.tune_mm_mlp_adapter = model_args.tune_mm_mlp_adapter
    model.config.tune_mm_vision_resampler = training_args.tune_mm_vision_resampler = model_args.tune_mm_vision_resampler
    if model_args.tune_mm_mlp_adapter or model_args.tune_mm_vision_resampler:
        model.requires_grad_(False)
    if model_args.tune_mm_mlp_adapter:
        for p in model.get_model().mm_projector.parameters():
            p.requires_grad = True
    if model_args.tune_mm_vision_resampler:
        for p in model.get_model().vision_resampler.parameters():
            p.requires_grad = True

    model.config.freeze_mm_mlp_adapter = training_args.freeze_mm_mlp_adapter
    if training_args.freeze_mm_mlp_adapter:
        for p in model.get_model().mm_projector.parameters():
            p.requires_grad = False

    model.config.freeze_mm_vision_resampler = training_args.freeze_mm_vision_resampler
    if training_args.freeze_mm_vision_resampler:
        for p in model.get_model().vision_resampler.parameters():
            p.requires_grad = False

    model.config.unfreeze_mm_vision_tower = model_args.unfreeze_mm_vision_tower
    if model_args.unfreeze_mm_vision_tower:
        vision_tower.requires_grad_(True)
    else:
        vision_tower.requires_grad_(False)

else:
    rank0_print(f"Using mm_tunable_parts: {model_args.mm_tunable_parts}")
    model.config.mm_tunable_parts = training_args.mm_tunable_parts = model_args.mm_tunable_parts
    # Set the entire model to not require gradients by default
    model.requires_grad_(False)
    vision_tower.requires_grad_(False)
    model.get_model().mm_projector.requires_grad_(False)
    model.get_model().vision_resampler.requires_grad_(False)
    # Parse the mm_tunable_parts to decide which parts to unfreeze
    tunable_parts = model_args.mm_tunable_parts.split(",")
    if "mm_mlp_adapter" in tunable_parts:
        for p in model.get_model().mm_projector.parameters():
            p.requires_grad = True
    if "mm_vision_resampler" in tunable_parts:
        for p in model.get_model().vision_resampler.parameters():
            p.requires_grad = True
    if "mm_vision_tower" in tunable_parts:
        for name, param in model.named_parameters():
            if "vision_tower" in name:
                param.requires_grad_(True)
    if "mm_language_model" in tunable_parts:
        for name, param in model.named_parameters():
            if "vision_tower" not in name and "mm_projector" not in name and "vision_resampler" not in name:
                param.requires_grad_(True)
```

After entering the else branch, it will first freeze all parts (including the injected LoRA), and then precisely unfreeze the specified parts according to mm_tunable_parts. When "mm_language_model" does not exist in mm_tunable_parts, LoRA will always be frozen; this is correct, after all, the backbone model is not trained. But then the problem comes: if "mm_language_model" exists in mm_tunable_parts, then it will enter the `if "mm_language_model" in tunable_parts:` sub-branch. At this time, it will unfreeze all parameters whose names do not contain vision_tower/mm_projector/vision_resampler, **including LoRA**! Then the bizarre thing comes: the LoRA parameters will be fine-tuned together with mm_language_model (i.e., the backbone model)! And we all know that the true meaning of LoRA is to freeze the backbone model and only fine-tune the LoRA parameters; doing this is obviously not the correct LoRA way!

#### Not using the mm_tunable_parts parameter

If you do not use the mm_tunable_parts parameter, it is indeed possible to achieve true LoRA fine-tuning. For example, set the mm_tunable_parts parameter to none, set the tune_mm_mlp_adapter parameter to False (why tune_mm_mlp_adapter is mentioned will be said later), and only set the lora_enable parameter to True. After my review, it will follow PEFT’s default behavior; generally it makes only the LoRA parameters trainable and freezes the backbone weights. At this time, it can indeed achieve true LoRA fine-tuning!

However, this fine-tuning method, besides the LoRA parameters, will freeze all other parts of the model (mm_mlp_adapter, vision_resampler, vision_tower), that is, it can only fine-tune LoRA. This is extremely, extremely inflexible! This is also absolutely not the result we want!

Furthermore, LLaVA-NeXT/scripts/train/finetune_ov.sh is an example fine-tuning script, in which the mm_tunable_parts parameter is clearly included. Ordinary researcher users will generally not, like me, read the code so carefully. When they want LoRA fine-tuning, generally they will directly set `--mm_tunable_parts="mm_mlp_adapter,mm_language_model"`, `--lora_enable True` (obviously the EgoBlind authors and issues#253 did it this way). They mistakenly think they used LoRA fine-tuning, but in fact not. In the end, compared to the previous full fine-tuning, they even fine-tuned the LoRA parameters additionally! VRAM usage and inference speed almost did not change; this is undoubtedly a misunderstanding and abuse of LoRA fine-tuning!

So I think the code in the `if "mm_language_model" in tunable_parts:` sub-branch in the above code block can be succinctly changed to the following code (I have already uploaded a Pull requests):

```python
if "mm_language_model" in tunable_parts:
    if training_args.lora_enable:
        rank0_print("LoRA enabled: Unfreezing ONLY LoRA parameters in LLM.")
        for name, param in model.named_parameters():
            if "lora_" in name:
                param.requires_grad_(True)
    else:
        rank0_print("LoRA disabled: Unfreezing FULL LLM backbone.")
        for name, param in model.named_parameters():
            if "vision_tower" not in name and "mm_projector" not in name and "vision_resampler" not in name and "mm_depth_projector" not in name:
                param.requires_grad_(True)
```

#### Test results

After my tests on the EgoBlind dataset, the training VRAM usage is about 56G after modification (originally close to 160G), the inference time is reduced from the original about 7 hours to 4.5 hours, fully in line with the expected effect of LoRA fine-tuning! The original fine-tuned model weight file was about 12GB, and the modified model weight file was about 2GB, also in line with the small-model expectation of LoRA fine-tuning! None of these phenomena fails to show that my analysis of the LoRA logic error in LLaVA-NeXT is correct! On the public EgoBlind dataset, the accuracy also increased from 57% to around 59%, and the effect improved to some extent!The modified code can perfectly adapt to the scripts in the repository; in fact, after I modified the code, it ran successfully in one go.
Sorry, I didn’t save the runtime time or VRAM usage data. Here I only attach comparison images of the model file size.

Image

The archive above is the model file obtained before the code change, with a size of 11.31 GB; the archive below is the model file obtained after the code change, with a size of 88.52 MB.
> When checking the code of this repository, I found that the tune_mm_mlp_adapter parameter is used to separately choose whether to fine-tune the MLP layer: the tune_mm_mlp_adapter parameter defaults to False, but if it is set to True, then except for the MLP layer, all other parts of the model will be frozen, including the LoRA parameters (even if lora_enable is set to True), that is, only the MLP layer is trained. [issues#438](https://github.com/LLaVA-VL/LLaVA-NeXT/issues/438)

### PR DESCRIPTION
**Pull request by lepelo0811**
Fixes #508 (https://github.com/LLaVA-VL/LLaVA-NeXT/issues/508)

### CODE REVIEW

**Review by kcz358 — DISMISSED** [review]

**Comment by lepelo0811** [comment]
Hi @Luodian and @kcz358,

Since @Nyquist24 accidentally closed PR #510, we synced up and decided to port all the exact changes from #510 directly into this PR (#509). 

We also made sure to include the additional code/fixes you provided in PR #511. So, the code in this PR (#509) is now completely identical to #511. However, since new changes have been pushed to this branch, we realize it will require a fresh review and approval from both of you before it can be finally merged.

Just a quick note regarding PR #512: @Nyquist24 opened that one before we had a chance to communicate. Using #512 as the merge target would mean starting the review process completely from scratch in a new thread, so updating #509 directly seemed like the cleanest and most efficient way forward.

We are truly sorry for all the back-and-forth and for taking up your valuable time with this confusion! Thank you so much for your patience, guidance, and all the hard work you put into maintaining this project. 

Wishing you both a wonderful rest of your week!

**Comment by Luodian** [comment]
Closing as superseded by merged PR #511.

Thank you to @lepelo0811 for reporting the bug, validating the behavior, and contributing the original concise fix. The merged solution keeps the original contributor credit path and includes the broader freeze fix plus the checkpoint/save follow-up needed for complete LoRA + adapter support.
